File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        num =0
        counts = []
        result = []
        curr = ""
        for ch in s:
            if ch.isdigit():
                num = num * 10+int(ch)
            elif ch == '[':
                counts.append(num)
                result.append(curr)
                num =0
                curr = ""
            elif ch ==']':
                repeat = counts[-1]
                counts.pop()
                temp = result[-1]
                result.pop()
                while(repeat !=0):
                    temp += curr
                    repeat -=1

                curr = temp
            else:
                curr += ch

        return curr

File: 0394-decode-string/0394-decode-string.java
class Solution {
    public String decodeString(String s) {
        Stack<Integer> counts = new Stack<>();
        Stack<String> result = new Stack<>();
        int num = 0;
        String curr = "";
        for(char c : s.toCharArray()){
            if(Character.isDigit(c)){
                num = num * 10 +(c - '0');
            }
            else if(c == '['){
                counts.push(num);
                result.push(curr);
                num =0;
                curr ="";
            }
            else if(c == ']'){
                int repeat = counts.peek();
                counts.pop();
                String temp = result.peek();
                result.pop();
                while(repeat != 0){
                    temp += curr;
                    repeat--;
                }
                curr = temp;
            }
            else {
                curr += c;
            }
        }
        return curr;
    }
}